sbi/inference/trainers/npe/npe_b.py

```diff
@@ -23,54 +23,80 @@ def __init__(
+        an importance-weighted log-loss. Unlike NPE-A the loss will make training
+        directly converge to the true posterior.
+        Thus, SNPE-B is not limited to Gaussian proposal.
 
-        raise NotImplementedError(
-            "SNPE-B is not yet implemented in the sbi package, see issue #199."
-        )
+        Args:
+            prior: A probability distribution that expresses prior knowledge about the
+                parameters, e.g. which ranges are meaningful for them.
+            density_estimator: If it is a string, use a pre-configured network of the
+                provided type (one of nsf, maf, mdn, made). Alternatively, a function
+                that builds a custom neural network can be provided. The function will
+                be called with the first batch of simulations (theta, x), which can
+                thus be used for shape inference and potentially for z-scoring. It
+                needs to return a PyTorch `nn.Module` implementing the density
+                estimator. The density estimator needs to provide the methods
+                `.log_prob` and `.sample()`.
+            device: Training device, e.g., "cpu", "cuda" or "cuda:{0, 1, ...}".
+            logging_level: Minimum severity of messages to log. One of the strings
+                INFO, WARNING, DEBUG, ERROR and CRITICAL.
+            summary_writer: A tensorboard `SummaryWriter` to control, among others, log
+                file location (default is `<current working directory>/logs`.)
+            show_progress_bars: Whether to show a progressbar during training.
+        """
 
         kwargs = del_entries(locals(), entries=("self", "__class__"))
         super().__init__(**kwargs)
 
     def _log_prob_proposal_posterior(
-        self, theta: Tensor, x: Tensor, masks: Tensor
+        self,
+        theta: Tensor,
+        x: Tensor,
+        masks: Tensor,
+        proposal: Optional[Any],
     ) -> Tensor:
         """
         Return importance-weighted log probability (Lueckmann, Goncalves et al., 2017).
 
         Args:
             theta: Batch of parameters θ.
             x: Batch of data.
-            masks: Whether to retrain with prior loss (for each prior sample).
+            masks: Mask that is True for prior samples in the batch in order to train
+                them with prior loss.
+            proposal: Proposal distribution.
 
         Returns:
-            Log probability of proposal posterior.
+            Importance-weighted log-probability of the proposal posterior.
         """
 
-        raise NotImplementedError
-
-        batch_size = theta.shape[0]
-
-        # Evaluate posterior.
-        log_prob_posterior = self._posterior.net.log_prob(theta, x)
-        log_prob_posterior = log_prob_posterior.reshape(batch_size)
-        utils.assert_all_finite(log_prob_posterior, "posterior eval")
+        # Evaluate prior
+        utils.assert_all_finite(self._prior.log_prob(theta), "prior eval")
```

Comment:
Good to check that prior log probs are reasonable, but a couple of thoughts:

1. If prior.log_prob returns -inf, this currently would cause an error to be raised. However, this just means that the importance weight for this sample is 0 (i.e the proposal tries to sample outside the prior) - this should not be an error. It does make sense to throw an error in the case that the prior log prob is nan or +inf.
2. I would make the message more descriptive, e.g. "prior log probs of proposal samples)